transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -243,14 +243,42 @@ public boolean writeOutbound(Object... msgs) {
      */
     public boolean finish() {
         close();
-        runPendingTasks();
+        checkException();
+        return isNotEmpty(inboundMessages) || isNotEmpty(outboundMessages);
+    }
 
+    private void finishPendingTasks() {
+        runPendingTasks();
         // Cancel all scheduled tasks that are left.
         loop.cancelScheduledTasks();
+    }
 
-        checkException();
+    @Override
+    public final ChannelFuture close() {
+        ChannelFuture future = super.close();
+        finishPendingTasks();
```

Comment:
No... We need to process them sfter as fireChannelInactive and fireChannelUnregistered are added as tasks via invokeLater.